### ISSUE TITLE
SQLSTATE[HY000]: [Microsoft][ODBC Driver 18 for SQL Server]Unicode conversion failed

### ISSUE THREAD

**Issue by ngochangngo52** [microsoft/msphpsql#1479]
Please check the [FAQ (frequently-asked questions)](https://github.com/Microsoft/msphpsql/wiki/FAQ) first. If you have other questions or something to report, please address the following (skipping questions might delay our responses):

**PHP version**  
8.1.23

**PHP SQLSRV or PDO_SQLSRV version**  
php-sqlsrv-5.11.1-1.el9.remi.8.1.x86_64

**Microsoft ODBC Driver version**  
[ODBC Driver 18 for SQL Server]

**SQL Server version**  

 Microsoft SQL Server 2016

**Client operating system**  
Alma Linux (WSL)

**Table schema**  

CREATE TABLE IF NOT EXISTS "OCRD" (
	"CardCode" NVARCHAR(15) NOT NULL COLLATE 'SQL_Latin1_General_CP850_CI_AS',
	"CardName" NVARCHAR(100) NULL DEFAULT NULL COLLATE 'SQL_Latin1_General_CP850_CI_AS',
	"CardType" CHAR(1) NULL DEFAULT NULL COLLATE 'SQL_Latin1_General_CP850_CI_AS',
	"GroupCode" SMALLINT NULL DEFAULT NULL,
	PRIMARY KEY ("CardCode")
);

**Problem description**  
I run simple query `select top 1 [CardType] from [dbo].[OCRD] where [CardCode] = 'MyClientCode'` and it raised error:

`SQLSTATE[HY000]: [Microsoft][ODBC Driver 18 for SQL Server]Unicode conversion failed`

I run in sqlcmd, it will have error:
`Sqlcmd: Error: Internal error at ReadAndHandleColumnData (Reason: Error reading column data). Unicode conversion failed`

**Expected behavior and actual behavior**  
Expected result: 'C'
Actual Result: Unicode conversion failed

**Repro code or steps to reproduce**
Just run `select top 1 [CardType] from [dbo].[OCRD] where [CardCode] = 'MyClientCode'` anywhere, it will throw error

**Comment by v-makouz** [comment]
What does the server give when running `select @@version`? I tried to reproduce this, but the table query gives errors on every SQL server I tried.

**Comment by ngochangngo52** [comment]
> select @@version

    Microsoft SQL Server 2016 (SP3-GDR) (KB5021129) - 13.0.6430.49 (X64) 
	Jan 22 2023 17:38:22 
	Copyright (c) Microsoft Corporation
	Standard Edition (64-bit) on Windows Server 2012 R2 Standard 6.3 <X64> (Build 9600: )

**Comment by v-makouz** [comment]
I can't seem to reproduce this one. I can execute the select query just fine with no errors. Also the table create query appears to be incorrect, the default values and NULL settings have to come after `COLLATE`, did it actually work as is?

**Comment by ngochangngo52** [comment]
> I can't seem to reproduce this one. I can execute the select query just fine with no errors. Also the table create query appears to be incorrect, the default values and NULL settings have to come after `COLLATE`, did it actually work as is?

It's OCRD table for SAP B1 database, you can find it here https://sap.erpref.com/?schema=BusinessOne9.3&module_id=3&table=OCRD
My app connected to existing SAP B1 database, I don't create new one.
Above creation SQL is just my extraction of structure using Heidi SQL so that can let you know the structure, I don't know the statement is weird to SQL Server :) It's sort of "reverse" from existing database, using Heidi SQL as tool.

The query can run in HeidiSQL, MS Azure Data Studio... but if I run in `SqlCmd` in the environment above, the error will throw.

**Comment by ngochangngo52** [comment]
> Also the table create query appears to be incorrect, the default values and NULL settings have to come after `COLLATE`, did it actually work as is?

I just updated the table creation, generate from MS Azure Data Studio should be correct one.

**Comment by ngochangngo52** [comment]
![image](https://github.com/microsoft/msphpsql/assets/107674191/115b31ea-1596-4915-8f4f-b703ad95c5d4)

**Comment by ngochangngo52** [comment]
Update:
the `sqlcmd` utility on Windows also work
![image](https://github.com/microsoft/msphpsql/assets/107674191/a058a606-8e3d-4ce8-ba37-824467f6c6ba)

The error is with the `sqlcmd` on Linux (Alma 9.2).

**Comment by ngochangngo52** [comment]
Just found this https://stackoverflow.com/a/77140830/223934
Basically installing glibc-gconv-extra solve my problem. 
`sudo dnf install glibc-gconv-extra`

But the problem doesn't exist if I don't update. Do you have any idea?

**Comment by v-makouz** [comment]
Probably that type of conversion isn't available by default on Alma Linux, I'm not sure since that's not an officially supported OS.

**Comment by ngochangngo52** [comment]
> Probably that type of conversion isn't available by default on Alma Linux, I'm not sure since that's not an officially supported OS.

Not only Alma Linux, but the entire **Fedora** family (including **RHEL**, **Oracle Linux**, **Centos**). For some distro, `glibc` already include `gconv`, but for **Fedora** family, the `gconv` has been split in to separate package `glibc-gconv-extra`
https://fedoraproject.org/wiki/Changes/Gconv_package_split_in_glibc

I think the ODBC should verify the dependency when installing.

**Comment by mtovmassian** [comment]
If installing *glibc-gconv-extra* on Red Hat x *glibc* systems solve the problem, does a solution exist on Alpine Linux x *musl* systems?

**Comment by v-maximko** [comment]
I don't believe that package is available on Alpine, but what encodings are you looking for?

**Comment by mtovmassian** [comment]
My question is related to this [stackoverflow question](https://stackoverflow.com/q/78370926/5472004) where the error could be a lack of enconding support for Japanese.
